### PR TITLE
HPCC-20397 Return better message when preflight failed

### PR DESCRIPTION
**Pull request by wangkx**
1. When a node is not responsive to preflight SSH command, SSH
returns 255 error code. CFRunSSH should report the "SSH failed
to connect 'ip'";
2. Run SSH command in foreground so that exception 127 can be
thrown when preflight script is not found.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
I create this PR to replace https://github.com/hpcc-systems/HPCC-Platform/pull/11860 which is targeted to 7.0.0. Please see the review comments on that PR. @mckellyln please review.

File: esp/services/ws_machine/ws_machineService.cpp [hunk 1]
```diff
@@ -1183,7 +1183,7 @@ int Cws_machineEx::runCommand(IEspContext& context, const char* sAddress, const
         IFRunSSH * connection = createFRunSSH();
         connection->init(command.str(),NULL,NULL,NULL,m_SSHConnectTimeoutSeconds,0);
         // executed as single connection
-        connection->exec(sAddress,NULL,true);
+        connection->exec(sAddress, NULL, false);
```
Comment (mckellyln) [inline esp/services/ws_machine/ws_machineService.cpp]: sAddress here is a const char *, but it appears the definition is:
```
    void exec(
              const IpAddress &ip,
              const char *workdirname,
              bool _background)
```
Where the first arg is a const IpAddress &

**Comment by mckellyln** [comment]
@wangkx looks good.  But have one question about exec() arg types.

File: esp/services/ws_machine/ws_machineService.cpp (hunk 1 above)
Comment (mckellyln) [reply esp/services/ws_machine/ws_machineService.cpp]: Talked with @jakesmith about this - we should add an 'explicit' to the IpAddress() ctor:
```
    explicit IpAddress(const char *text)                         { ipset(text); }
```
In jsocket.hpp to make this more correct.  And then we can no longer pass in sAddress here, but need to first create an IpAddress type and init with the sAddress text and pass that in.

Comment (jakesmith) [reply esp/services/ws_machine/ws_machineService.cpp]: Thanks @mckellyln 

I think this PR is good to merge now.

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
yes, ok - @richardkchapman ok to merge.

File: esp/services/ws_machine/ws_machineService.cpp (hunk 1 above)
Comment (richardkchapman) [inline esp/services/ws_machine/ws_machineService.cpp]: What will be the impact of changing the "background" parameter to false? Does it mean multiple opens do not happen in parallel, for example? That could be an issue on a large cluster

**Comment by richardkchapman** [comment]
@wangkx One question regarding performance

File: esp/services/ws_machine/ws_machineService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_machine/ws_machineService.cpp]: If a user selects HPCC components from multiple IPs, the ws_machine code will create one thread for each IP and call this runCommand() (and this exec()) in each thread. So, the multiple exec()'s will be run in parallel.

**Comment by mckellyln** [comment]
@richardkchapman you ok with answer ?

File: esp/services/ws_machine/ws_machineService.cpp (hunk 1 above)
Comment (richardkchapman) [reply esp/services/ws_machine/ws_machineService.cpp]: So why did it say true before?

**Comment by richardkchapman** [comment]
@wangkx One more question, just to check

File: esp/services/ws_machine/ws_machineService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_machine/ws_machineService.cpp]: This is my guessing: the original code (https://github.com/hpcc-systems/HPCC-Platform/commit/993519e6030832340948813816945401fedfc2a5) does not realize that there is already multi-thread implementation.

Comment (mckellyln) [reply esp/services/ws_machine/ws_machineService.cpp]: I think invokeProgram() would run in the foreground ??

Comment (mckellyln) [reply esp/services/ws_machine/ws_machineService.cpp]: I think? the exec() code runs the command and waits for it to finish and collects the output, so not really understanding why it was originally set to background when exit code and output of command is wanted.
@wangkx could we test with >1 IP and a long running (ie adding sleep 30) preflight to make sure it does not take longer than with 1 IP ?

Comment (mckellyln) [reply esp/services/ws_machine/ws_machineService.cpp]: @richardkchapman fyi

**Comment by richardkchapman** [comment]
@mckellyln Are you happy with this?

**Comment by mckellyln** [comment]
@richardkchapman yes, I am ok to merge.

**Review by mckellyln — APPROVED** [review]

**Pull request merged**